closed/src/java.base/share/classes/com/sun/crypto/provider/NativeChaCha20Cipher.java

```diff
@@ -857,6 +857,31 @@ private static byte[] intToLittleEndian (long i) {
         return ret;
     }
 
+    private void ChaCha20Init(boolean modeChanged) {
+        // Optimize the initialization of chacha20-poly1305 when they have the same ossl_mode
+        // otherwise, we have to reinitialize based upon the ossl_mode.
+        int ret = nativeCrypto.ChaCha20Init(context, ossl_mode, openssl_iv, openssl_iv.length, keyBytes, keyBytes.length, !modeChanged);
+
+        if (ret == -1) {
+            throw new ProviderException("Error in Native ChaCha20Cipher");
+        }
+        opensslInitialized = true;
+    }
+
+    /**
+     * Reset the Cipher's state to the appropriate values.
+     *
+     * Note: The cipher's internal "initialized" field is set differently
+     * for ENCRYPT_MODE and DECRYPT_MODE in order to allow DECRYPT_MODE
+     * ciphers to reuse the key/nonce/counter values.  This kind of reuse
+     * is disallowed in ENCRYPT_MODE.
+     */
+    private void resetStartState() {
+        aadDone = false;
+        initialized = (direction == Cipher.DECRYPT_MODE);
```

Comment:
I think this should only (potentially) clear `initialized`. If it's already set, leave it that way:
```java
        if (direction != Cipher.DECRYPT_MODE) {
            initialized = false;
        }
```

Reply:
Changed.